### PR TITLE
Cleanup mdivide_* and tests

### LINKED ISSUE #1714: remove uses of promote_common from mdivide_* functions and improve tests
## Description
Several of the `mdivide_*` set of functions use `promote_common` when instead `return_type_t` would be much cleaner.

Tests for these functions are inconsistent with respect to the left/right versions, meaning that some corner cases are covered for one function but not the other. Moreover, `test/unit/math/mix/fun/mdivide_left_test.cpp` somehow believes that the input matrix should be symmetric, while instead there's no such requirement.

## Example
In `mdivide_left_tri` there's this code:
```
return promote_common<Eigen::Matrix<T1, R1, C1>, Eigen::Matrix<T2, R1, C1> >(
           A)
     .template triangularView<TriView>()
    .solve(
        promote_common<Eigen::Matrix<T1, R2, C2>, Eigen::Matrix<T2, R2, C2> >(
            b));
```
which can be rewritten as:
```
return Eigen::Matrix<return_type_t<T1, T2>, R1, C1>(A)
    .template triangularView<TriView>()
    .solve(Eigen::Matrix<return_type_t<T1, T2>, R2, C2>(b));
```
## Expected Output
Same as before.

#### Current Version:
v3.1.0

### PR DESCRIPTION
**Pull request by mcol**
## Summary

This cleans up the `mdivide` functions, mainly to use `return_type_t` instead of `promote_common`. There are a couple of unused header removed and a correction to the return type declared for the fwd version of `mdivide_left_tri_low`. Fixes #1714.

## Tests
Reworked and expanded so that the left and right versions contain the same tests. Some boundary condition tests will be completed in #1689.

## Side Effects

None.

## Checklist

- [X] Math issue #1714

- [X] Copyright holder: Marco Colombo

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [X] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [X] the code is written in idiomatic C++ and changes are documented in the doxygen

- [X] the new changes are tested

### CODE REVIEW

**Comment by mcol** [comment]
I have two more PRs I'm preparing that depend on this. @rok-cesnovar Would you mind review this?

**Comment by rok-cesnovar** [comment]
Sure. I can do this in a few hours when I will be at my computer.

File: stan/math/fwd/fun/mdivide_left_tri_low.hpp [hunk 1]
```diff
@@ -13,7 +13,7 @@ namespace stan {
 namespace math {
 
 template <typename T, int R1, int C1, int R2, int C2>
-inline Eigen::Matrix<fvar<T>, R1, C1> mdivide_left_tri_low(
+inline Eigen::Matrix<fvar<T>, R1, C2> mdivide_left_tri_low(
```
Comment (rok-cesnovar) [inline stan/math/fwd/fun/mdivide_left_tri_low.hpp]: Nice catch!

File: stan/math/prim/fun/mdivide_left.hpp [hunk 2]
```diff
@@ -31,12 +31,9 @@ inline Eigen::Matrix<return_type_t<T1, T2>, R1, C2> mdivide_left(
...
   check_multiplicable("mdivide_left", "A", A, "b", b);
-  return promote_common<Eigen::Matrix<T1, R1, C1>, Eigen::Matrix<T2, R1, C1> >(
-             A)
-      .lu()
-      .solve(
-          promote_common<Eigen::Matrix<T1, R2, C2>, Eigen::Matrix<T2, R2, C2> >(
-              b));
+
+  return Eigen::Matrix<return_type_t<T1, T2>, R1, C1>(A).lu().solve(
+      Eigen::Matrix<return_type_t<T1, T2>, R2, C2>(b));
```
Comment (rok-cesnovar) [inline stan/math/prim/fun/mdivide_left.hpp]: Much nicer!

**Comment by mcol** [comment]
Thank you!
